New Resource : `azurerm_arc_resource_bridge_appliance`

sinbai opened the pull request:
Swagger: https://github.com/Azure/azure-rest-api-specs/blob/main/specification/resourceconnector/resource-manager/Microsoft.ResourceConnector/stable/2022-10-27/appliances.json

Test results:
```
PASS: TestAccArcResourceBridgeAppliance_basic (221.73s)
PASS: TestAccArcResourceBridgeAppliance_update (343.70s)
PASS: TestAccArcResourceBridgeAppliance_complete (243.17s)
PASS: TestAccArcResourceBridgeAppliance_requiresImport (218.86s)
```
Note: Need to run `az provider register --namespace Microsoft.ResourceConnector` to register provider before running tests.

tombuildsstuff reviewed (changes requested):
hi @sinbai 

Thanks for this PR - I've taken a look through and left some comments inline, if those can be addressed then we should be able to take another look here.

It's worth noting that when a field has to be set via the Update method, that we can first Create the resource and then use the Update API to update this value - as such we don't need to ask users to conditionally apply their Terraform Configuration in two stages.

It's also worth noting there's conflation between the Appliance and the Resource Connector Appliance in the documentation, presumably the Appliance has to exist within AKS/VMware etc first? If so, then we're also missing test cases showing this working end-to-end - could we clarify the behaviour of this service to make it fully usable here?

Thanks!

sinbai commented on internal/services/resourceconnector/resourceconnector_appliance_resource.go:
```diff
@@ -0,0 +1,302 @@
+
+			if model.Version != "" {
+				return fmt.Errorf("the version cannot be set when creating %s. it could be set from upgrade call", id)
+			}
```
For `PublicKey` property, I fixed it as the suggested. For the `Version` property, I have removed it as it is not a user settable property after confirming with service team.

sinbai commented on internal/services/resourceconnector/resourceconnector_appliance_resource.go:
```diff
@@ -0,0 +1,302 @@
+			}
+
+			parameters := appliances.Appliance{
+				Location: model.Location,
```
Fixed.

sinbai commented on internal/services/resourceconnector/resourceconnector_appliance_resource.go:
```diff
@@ -0,0 +1,302 @@
+				return fmt.Errorf("expanding SystemAssigned Identity: %+v", err)
+			}
+
+			parameters.Identity = identity
```
Fixed.

sinbai commented on internal/services/resourceconnector/resourceconnector_appliance_resource.go:
```diff
@@ -0,0 +1,302 @@
+
+			if metadata.ResourceData.HasChange("version") {
+				parameters.Properties.Version = pointer.To(model.Version)
+			}
```
Fixed.

sinbai commented on internal/services/resourceconnector/resourceconnector_appliance_resource.go:
```diff
@@ -0,0 +1,302 @@
+				return fmt.Errorf("expanding SystemAssigned Identity: %+v", err)
+			}
+
+			parameters.Identity = identity
```
Fixed.

sinbai commented on internal/services/resourceconnector/resourceconnector_appliance_resource.go:
```diff
@@ -0,0 +1,302 @@
+				if p := v.InfrastructureConfig; p != nil {
+					state.Provider = pointer.From(p.Provider)
+				}
+			}
```
Fixed.

sinbai commented on internal/services/resourceconnector/resourceconnector_appliance_resource.go:
```diff
@@ -0,0 +1,302 @@
+				if old.(string) != "" {
+					return fmt.Errorf("the public_key can not be updated once it is set")
+				}
+			}
```
Fixed.

sinbai commented on internal/services/resourceconnector/resourceconnector_appliance_resource_test.go:
```diff
@@ -0,0 +1,187 @@
+
+func TestAccResourceConnectorAppliance_update(t *testing.T) {
+	data := acceptance.BuildTestData(t, "azurerm_resource_connector_appliance", "test")
+	l := ResourceConnectorApplianceResource{}
```
Fixed.

sinbai commented on internal/services/resourceconnector/resourceconnector_appliance_resource_test.go:
```diff
@@ -0,0 +1,187 @@
+
+func TestAccResourceConnectorAppliance_basic(t *testing.T) {
+	data := acceptance.BuildTestData(t, "azurerm_resource_connector_appliance", "test")
+	l := ResourceConnectorApplianceResource{}
```
Fixed.

sinbai commented on internal/services/resourceconnector/resourceconnector_appliance_resource_test.go:
```diff
@@ -0,0 +1,187 @@
+
+func TestAccResourceConnectorAppliance_complete(t *testing.T) {
+	data := acceptance.BuildTestData(t, "azurerm_resource_connector_appliance", "test")
+	l := ResourceConnectorApplianceResource{}
```
Fixed.

sinbai commented on internal/services/resourceconnector/resourceconnector_appliance_resource_test.go:
```diff
@@ -0,0 +1,187 @@
+
+func TestAccResourceConnectorAppliance_requiresImport(t *testing.T) {
+	data := acceptance.BuildTestData(t, "azurerm_resource_connector_appliance", "test")
+	l := ResourceConnectorApplianceResource{}
```
Fixed.

sinbai commented on internal/services/resourceconnector/resourceconnector_appliance_resource_test.go:
```diff
@@ -0,0 +1,187 @@
+  features {}
+}
+resource "azurerm_resource_group" "test" {
+  name     = "acctestRG-Appliances-%[1]d"
```
Fixed.

sinbai commented on internal/services/resourceconnector/resourceconnector_appliance_resource_test.go:
```diff
@@ -0,0 +1,187 @@
+  location = "%[2]s"
+}
+resource "azurerm_resource_connector_appliance" "test" {
+  name                    = "acctestAppliances-%[1]d"
```
Fixed.

sinbai commented on website/docs/r/resource_connector_appliance.html.markdown:
```diff
@@ -0,0 +1,92 @@
+
+Manages an Appliances.
+
+-> **Note:** Before using this resource, it's required to submit the request of registering the providers with Azure CLI `az provider register --namespace Microsoft.ResourceConnector`.
```
Fixed.

sinbai commented on website/docs/r/resource_connector_appliance.html.markdown:
```diff
@@ -0,0 +1,92 @@
+layout: "azurerm"
+page_title: "Azure Resource Manager: azurerm_resource_connector_appliance"
+description: |-
+  Manages an Appliance.
```
Just the connection. The description has been updated.

sinbai commented on website/docs/r/resource_connector_appliance.html.markdown:
```diff
@@ -0,0 +1,92 @@
+
+# azurerm_resource_connector_appliance
+
+Manages an Appliances.
```
Just the connection. It has been updated.

sinbai commented on website/docs/r/resource_connector_appliance.html.markdown:
````diff
@@ -0,0 +1,92 @@
+```hcl
+resource "azurerm_resource_group" "example" {
+  name     = "example"
+  location = "eastus"
````
Fixed.

sinbai commented on website/docs/r/resource_connector_appliance.html.markdown:
```diff
@@ -0,0 +1,92 @@
+
+* `location` - (Required) The Azure Region where the Appliance should exist. Changing this forces a new resource to be created.
+
+* `distro` - (Required) Specifies the resource group where the resource exists. The possible value is `AKSEdge`.
```
Fixed.

sinbai commented on website/docs/r/resource_connector_appliance.html.markdown:
```diff
@@ -0,0 +1,92 @@
+
+The following arguments are supported:
+
+* `name` - (Required) The Name which should be used for this Appliance. Changing this forces a new resource to be created.
```
Fixed.

sinbai commented on website/docs/r/resource_connector_appliance.html.markdown:
```diff
@@ -0,0 +1,92 @@
+
+* `infrastructure_provider`- (Required) The infrastructure provider about the connected Appliance. Possible values are `HCI`,`SCVMM` and `VMWare`. Changing this forces a new resource to be created.
+
+* `public_key`- (Required) The `public_key` is an RSA public key in PKCS1 format encoded in base64.
```
Fixed.

sinbai commented on website/docs/r/resource_connector_appliance.html.markdown:
```diff
@@ -0,0 +1,92 @@
+
+* `resource_group_name` - (Required) Specifies the resource group where the resource exists. Changing this forces a new resource to be created.
+
+* `location` - (Required) The Azure Region where the Appliance should exist. Changing this forces a new resource to be created.
```
Resource Connector Appliance, fixed.

sinbai commented on website/docs/r/resource_connector_appliance.html.markdown:
```diff
@@ -0,0 +1,92 @@
+
+* `name` - (Required) The Name which should be used for this Appliance. Changing this forces a new resource to be created.
+
+* `resource_group_name` - (Required) Specifies the resource group where the resource exists. Changing this forces a new resource to be created.
```
Fixed.

sinbai commented on website/docs/r/resource_connector_appliance.html.markdown:
```diff
@@ -0,0 +1,92 @@
+
+* `public_key`- (Required) The `public_key` is an RSA public key in PKCS1 format encoded in base64.
+
+* `tags` - (Optional) A mapping of tags which should be assigned to the Appliance.
```
Resource Connector Appliance, fixed.

sinbai commented on internal/services/resourceconnector/resourceconnector_appliance_resource.go:
```diff
@@ -0,0 +1,302 @@
+			Optional: true,
+			ValidateFunc: validation.StringInSlice([]string{
+				"latest",
+			}, false),
```
After confirming with service team, the `version` is not a user settable property, I have removed it.

sinbai commented on website/docs/r/resource_connector_appliance.html.markdown:
```diff
@@ -0,0 +1,92 @@
+
+* `version` - (Optional) The version of the Appliance. Possible values is `latest`.
+
+-> **NOTE:** The `version` cannot be populated apart from upgrade call.
```
After confirming with service team, the version is not a user settable property, I have removed it.

sinbai commented on website/docs/r/resource_connector_appliance.html.markdown:
````diff
@@ -0,0 +1,92 @@
+Appliance can be imported using the `resource id`, e.g.
+
+```shell
+terraform import azurerm_resource_connector_appliance.example /subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/providers/Microsoft.ResourceConnector/apliances/appliancesExample
````
appliances, fixed.

sinbai commented on website/docs/r/resource_connector_appliance.html.markdown:
```diff
@@ -0,0 +1,92 @@
+* `create` - (Defaults to 60 minutes) Used when creating the Appliance.
+* `read` - (Defaults to 5 minutes) Used when retrieving the Appliance.
+* `update` - (Defaults to 30 minutes) Used when updating the Appliance.
+* `delete` - (Defaults to 30 minutes) Used when deleting the Appliance.
```
Resource Connector Appliance, fixed.

sinbai commented on website/docs/r/resource_connector_appliance.html.markdown:
```diff
@@ -0,0 +1,92 @@
+
+In addition to the Arguments listed above - the following Attributes are exported: 
+
+* `id` - The ID of the Appliance.
```
Resource Connector Appliance, fixed.

sinbai commented on website/docs/r/resource_connector_appliance.html.markdown:
```diff
@@ -0,0 +1,92 @@
+
+* `tags` - (Optional) A mapping of tags which should be assigned to the Appliance.
+
+* `version` - (Optional) The version of the Appliance. Possible values is `latest`.
```
Resource Connector Appliance, fixed.

sinbai commented:
> hi @sinbai
> 
> Thanks for this PR - I've taken a look through and left some comments inline, if those can be addressed then we should be able to take another look here.
> 
> It's worth noting that when a field has to be set via the Update method, that we can first Create the resource and then use the Update API to update this value - as such we don't need to ask users to conditionally apply their Terraform Configuration in two stages.
> 
> It's also worth noting there's conflation between the Appliance and the Resource Connector Appliance in the documentation, presumably the Appliance has to exist within AKS/VMware etc first? If so, then we're also missing test cases showing this working end-to-end - could we clarify the behaviour of this service to make it fully usable here?
> 
> Thanks!

@tombuildsstuff  thanks for your detailed feedback. I have fixed the inline feedback. Could you please take another look?

For above `test cases showing this working end-to-end`, the following is the feedback I got from the service team.
```
Appliance RP is a hybrid resource provider service. We have a few operations that deploy a cluster on prem for customers like Validate, Prepare, Deploy are run on prem only and can be done via Arc Appliance CLI as this requires specific parameters depending on the on prem stack being used to deploy the cluster on. 

We do create an Azure resource to help link the on prem resource to Azure, and that is done using  RP operations (that we support using various SDKs) like Create, Update, Get, List, etc, for which we are also looking to support Terraform as the RP is GAing.
```

> Link to public doc - [Azure Arc resource bridge (preview) overview - Azure Arc | Microsoft Learn](https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Flearn.microsoft.com%2Fen-us%2Fazure%2Fazure-arc%2Fresource-bridge%2Foverview&data=05%7C01%7Cv-elenaxin%40microsoft.com%7Cc0f9934bf9d24dab02a408dbb31b47b5%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C638300702068993345%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=frjs4LxVjoCP8z76XX%2Fxs60t0oU6IuB8%2B08WC6GC5Vo%3D&reserved=0)

Per above feedback, I assume that we could not add end-to-end testing in TF, what do you think?

stephybun reviewed (changes requested):
Thanks for making those changes @sinbai.

There are a few things that still need to be fixed.

Also since this is from the Azure Arc product line this resource should be renamed to `azurerm_arc_bridge_appliance`/`azurerm_arc_bridge_connector_appliance` or something of the sort? This is because we should go with names that are familiar to the users, in this case since this is being marketed as Arc Bridge. Also existing arc resources within the provider like the Arc Kubernetes cluster or Arc Virtual Machine are also named as such.

sinbai commented on internal/services/resourceconnector/resourceconnector_appliance_resource.go:
```diff
@@ -0,0 +1,275 @@
+	Identity          []identity.ModelSystemAssigned `tfschema:"identity"`
+	Provider          appliances.Provider            `tfschema:"infrastructure_provider"`
+	PublicKey         string                         `tfschema:"public_key"`
+	Tags              map[string]string              `tfschema:"tags"`
```
Fixed.

sinbai commented on internal/services/resourceconnector/resourceconnector_appliance_resource.go:
```diff
@@ -0,0 +1,275 @@
+						Provider: pointer.To(model.Provider),
+					},
+				},
+				Tags: pointer.To(model.Tags),
```
Fixed.

sinbai commented on internal/services/resourceconnector/resourceconnector_appliance_resource.go:
```diff
@@ -0,0 +1,275 @@
+			}
+
+			if metadata.ResourceData.HasChange("tags") {
+				parameters.Tags = pointer.To(model.Tags)
```
Fixed.

sinbai commented on internal/services/resourceconnector/resourceconnector_appliance_resource.go:
```diff
@@ -0,0 +1,275 @@
+			if model := resp.Model; model != nil {
+				state.Location = location.Normalize(model.Location)
+				state.Identity = identity.FlattenSystemAssignedToModel(model.Identity)
+				state.Tags = pointer.From(resp.Model.Tags)
```
Fixed.

sinbai commented on website/docs/r/resource_connector_appliance.html.markdown:
```diff
@@ -0,0 +1,86 @@
+
+* `infrastructure_provider`- (Required) The infrastructure provider about the connected Resource Connector Appliance. Possible values are `HCI`,`SCVMM` and `VMWare`. Changing this forces a new resource to be created.
+
+* `public_key`- (Optional) The `public_key` is an RSA public key in PKCS1 format encoded in base64. Changing this forces a new resource to be created.
```
Thanks for your explanation, fixed:)

sinbai commented:
> Thanks for making those changes @sinbai.
> 
> There are a few things that still need to be fixed.
> 
> Also since this is from the Azure Arc product line this resource should be renamed to `azurerm_arc_bridge_appliance`/`azurerm_arc_bridge_connector_appliance` or something of the sort? This is because we should go with names that are familiar to the users, in this case since this is being marketed as Arc Bridge. Also existing arc resources within the provider like the Arc Kubernetes cluster or Arc Virtual Machine are also named as such.

@stephybun thanks for you detailed feedback. I updated the resource name to `azurerm_arc_resource_bridge_appliance` based on [What is Azure Arc resource bridge (preview)](https://learn.microsoft.com/en-us/azure/azure-arc/resource-bridge/overview), WDYT? Also, I updated the other comments, could you please take another look?

stephybun reviewed (approved):
Thanks @sinbai LGTM 🍰
